Enable reentrant dispatch for decompositions

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81677
* __->__ #81598

This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

davidberard98 committed:
Enable reentrant dispatch for decompositions

This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -403,7 +421,7 @@ def wrap_fake(x):
         if use_fake:  # type: ignore[attr-defined]
             args = pytree.tree_map(wrap_fake, args)
 
-        with decompose(decomposition_table), fake_tensor_mode, proxy_mode:  # type: ignore[attr-defined]
+        with decompose(decomposition_table), fake_tensor_mode, proxy_mode, proxy_dispatch_mode_ctx(proxy_mode):  # type: ignore[attr-defined] # noqa: B950
```
is there a better way to handle this?

- if I use a backslash for line continuation, there's no way to apply `# type: ignore[attr-defined]` to suppress mypy error
- `with (a, b, c, d):` allows splitting the line to <140 char/line and `#type: ignore[attr-defined]` works with it, but it only works in python 3.9+
- `contextlib.nested` is deprecated

ezyang replied:
you could use ExitStack if you like

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -75,7 +90,10 @@ def proxy_call(func_overload, args, kwargs=None):
     if func_overload in CURRENT_DECOMPOSITION_TABLE:
-        return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
+        global PROXY_DISPATCH_MODE
+        proxy_mode = enable_torch_dispatch_mode(PROXY_DISPATCH_MODE) if PROXY_DISPATCH_MODE else nullcontext()
```
You don't need to do this. `proxy_call` is called from `__torch_dispatch__` on the ProxyMode object you're looking for. Pass it in as an argument.

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

davidberard98 committed:
Update on "Enable reentrant dispatch for decompositions"


This allows us to avoid tracing through CompositeImplicitAutograd ops
when decomposing via make_fx or other decomposition methods that use
tracing.

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -70,12 +70,14 @@ def maybe_disable_fake_tensor_mode():
     if func_overload in CURRENT_DECOMPOSITION_TABLE:
-        return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
+        proxy_mode = enable_torch_dispatch_mode(dispatch_mode) if dispatch_mode else nullcontext()
+        with proxy_mode, torch.overrides.enable_reentrant_dispatch():
```
Now I'm a little puzzled... how come we need `enable_reentrant_dispatch`? Based on what you've told us, it seems to be required, but I'm blanking on exactly why mechanically it's needed. In particular, if there is a dispatch mode stack, we don't seem to require people to reenable reentrant dispatch when they call into func (which will ostensibly go to the next dispatch mode in the stack).

ezyang replied:
This doesn't smell right; either there's another bug lurking, or we need to turn these two context managers into a consolidated context manager to make it easier for people to do this correctly.

davidberard98 replied:
I think we want to use `enable_reentrant_dispatch` so that we can reset the higher dispatch bits, in order to trace through CompositeImplicitAutograd ops. (On the first dispatch, PythonTLSSnapshot saves the original DispatchKeySet so that `enable_reentrant_dispatch` can restore the same DispatchKeySet).

re: turning these two context managers into a consolidated context manager - we could make `enable_reentrant_dispatch` take an optional torch_dispatch object? But there are also cases like this: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/composite_compliance.py#L200-L205 where `enable_reentrant_dispatch()` is sometimes getting called via `enable_torch_dispatch_mode()` on a tensor subclass, where it might be difficult to figure out which torch_dispatch mode to use. (see failures on #81677)

ezyang replied:
OK I take it back, what's going on here is probably wrong and we should find some other mechanism for this.

If you are using `enable_reentrant_dispatch` to shunt as back to CompositeImplicitAutograd, not only are we reenabling the composites, but we're also reenabling autograd, which can lead to weird things where we start building up grad fn, even though at the outer layer we've already set autograd. Let me patch your change in and see if I can figure out how to do it better.

ezyang replied:
OK, this is probably the easiest way to solve this for now.

Long term what would make more sense is for proxy tensor to be able to directly make use of CompositeImplicitAutograd registered kernels *in its body itself*. The same way that when a CPU kernel calls a composite kernel, that will hit the composite registration in CPU itself (without reenabling reentrant dispatch) and eventually getting to the real CPU backend kernels. Also the same way we consult a decomposition table here. We just have no convenient way to actually get at these registrations at the moment; @anjali411's proposal to get out ops from the dispatch table would help a little, but we would need to be able to query for composites specifically.

Changing https://github.com/pytorch/pytorch/issues/81682 would also help, as the tracing kernel would then be a fallback kernel and defer to composites.

ezyang replied:
Another way of fixing this, by @albanD, is add a composite key separate from autograd which doesn't get disabled when we go in. cc @zou3519

ezyang reviewed (changes requested):
comment coming soon

davidberard98 commented:
@pytorchbot merge

jansel commented:
This PR seems to be breaking AOT autograd.  A lot of models started failing with it.

One example:
```
cuda train resnet18                           ERROR:root:unhandled error
Traceback (most recent call last):
  File "/home/jansel/torchdynamo/benchmarks/common.py", line 702, in run_one_model
    new_result = model_iter_fn(model, example_inputs)
  File "./benchmarks/torchbench.py", line 355, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 355, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 359, in forward_and_backward_pass
    pred = mod(*cloned_inputs)
  File "/home/jansel/pytorch/torch/nn/modules/module.py", line 1186, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/jansel/torchvision/torchvision/models/resnet.py", line 284, in forward
    def forward(self, x: Tensor) -> Tensor:
  File "/home/jansel/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
    return fn(*args, **kwargs)
  File "/home/jansel/functorch/functorch/_src/aot_autograd.py", line 652, in forward
    return compiled_f(
  File "/home/jansel/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
    return fn(*args, **kwargs)
  File "/home/jansel/functorch/functorch/_src/aot_autograd.py", line 195, in forward
    fx_g = make_fx(joint_forward_backward, aot_decompositions)(
  File "/home/jansel/pytorch/torch/fx/experimental/proxy_tensor.py", line 409, in wrapped
    t = dispatch_trace(wrap_key(f, args), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/home/jansel/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
    return fn(*args, **kwargs)
  File "/home/jansel/pytorch/torch/fx/experimental/proxy_tensor.py", line 248, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/home/jansel/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
    return fn(*args, **kwargs)
  File "/home/jansel/pytorch/torch/fx/_symbolic_trace.py", line 714, in trace
    (self.create_arg(fn(*args)),),
  File "/home/jansel/pytorch/torch/fx/_symbolic_trace.py", line 549, in flatten_fn
    tree_out = root_fn(*tree_args)
  File "/home/jansel/pytorch/torch/fx/experimental/proxy_tensor.py", line 272, in wrapped
    out = f(*tree_args)
  File "/home/jansel/functorch/functorch/_src/aot_autograd.py", line 98, in joint_forward_backward
    backward_out = torch.autograd.grad(
  File "/home/jansel/pytorch/torch/autograd/__init__.py", line 294, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.HalfTensor [512]], which is output 0 of torch::autograd::CopyBackwards, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).
```

If I revert this PR it passes.

You can repro in dynamo with:
```
./benchmarks/torchbench.py -dcuda --float16 --inductor --training -k resnet18
```


cc @Chillee

Chillee commented:
@jansel Yeah we plan on reverting this + fixing this issue another way.

cc: @ezyang

ezyang commented:
@pytorchbot revert

ezyang commented:
@pytorchbot -m "out of tree failures" -c nosignal

ezyang commented:
@pytorchbot revert -m "out of tree failures" -c nosignal

xwang233 commented:
Hi @davidberard98 , we saw that this PR introduced many regression in AOT_autograd + nvfuser with **AMP** modes. Now that the PR was reverted and performances were recovered, I think it's still worth to check those AMP models before relanding a similar PR. FP32 mode seems unaffected.

Repro steps:

Use benchmark.py script from TIMM https://github.com/rwightman/pytorch-image-models/blob/master/benchmark.py

```
python benchmark.py --model resnet50 -b 128 --img-size 224 --bench train --aot-autograd --fuser nvfuser --amp
```

On V100

```
Good:

WARNING: Overriding precision to 'amp' since --amp flag set.
Benchmarking in amp precision. NCHW layout. torchscript disabled
Model resnet50 created, param count: 25557032
Running train benchmark on resnet50 for 40 steps w/ input size (3, 224, 224) and batch size 128.
Train [8/40]. 935.47 samples/sec. 136.829 ms/step.
Train [16/40]. 966.76 samples/sec. 132.401 ms/step.
Train [24/40]. 977.61 samples/sec. 130.931 ms/step.
Train [32/40]. 982.99 samples/sec. 130.215 ms/step.
Train [40/40]. 986.23 samples/sec. 129.787 ms/step.
Train benchmark of resnet50 done. 975.90 samples/sec, 129.79 ms/sample
--result
{
    "model": "resnet50",
    "train_samples_per_sec": 975.9,
    "train_step_time": 129.787,
    "train_batch_size": 128,
    "train_img_size": 224,
    "param_count": 25.56
}

Bad:

Benchmarking in amp precision. NCHW layout. torchscript disabled
Model resnet50 created, param count: 25557032
Running train benchmark on resnet50 for 40 steps w/ input size (3, 224, 224) and batch size 128.
Train [8/40]. 726.97 samples/sec. 176.072 ms/step.
Train [16/40]. 727.29 samples/sec. 175.996 ms/step.
Train [24/40]. 727.38 samples/sec. 175.974 ms/step.
Train [32/40]. 727.42 samples/sec. 175.965 ms/step.
Train [40/40]. 727.48 samples/sec. 175.949 ms/step.
Train benchmark of resnet50 done. 723.72 samples/sec, 175.95 ms/sample
--result
{
    "model": "resnet50",
    "train_samples_per_sec": 723.72,
    "train_step_time": 175.949,
    "train_batch_size": 128,
    "train_img_size": 224,
    "param_count": 25.56
}
```

cc @ptrblck

Chillee commented:
@xwang233 This is probably since this change induces some slightly different decompositions of operators, which are probably prohibiting some fusions.

It should be fairly easy fixes.

cc: @ezyang

ezyang commented:
@xwang233 can you give us before and after straight line operator dumps of the model post lowering? I can show you how to do it if you do not know how.

xwang233 commented:
@ezyang , I'm not sure how to do that. 😅

Chillee commented:
@xwang233 I think `PYTORCH_JIT_LOG_LEVEL="graph_fuser"` would be sufficient (i.e. NVFuser fusion group debug output).

xwang233 commented:
I posted the results at https://gist.github.com/xwang233/6f7cb7b27e47b92b239bd3e1e26c5e8e

davidberard98 commented:
@xwang233 actually can you do it with `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser"`, I think we need the `>>` to increase verbosity and get the debug of the fusion groups

xwang233 commented:
The repro step in https://github.com/pytorch/pytorch/pull/81598#issuecomment-1196069402 is a one-liner and straightforward. Would you be able to reproduce the regression in your environment?
